[DevTools] - Highlight rendered by elements on hover.

hristo-kanchev opened the pull request:
##  Summary

Resolve: #18472 

What's been done:
- Extracted `highlightNativeElement` functionality in its own hook in order to reuse it in multiple components.
- Replaced old highlight code in `Tree` view with the `useHighlightNativeElement` hook.
- Added support for highlighting native elements in the `rendered by` section.
- Added hover state to `rendered by` elements for better visibility on what's being hovered - same as in `Tree` view. (@bvaughn I can revert this part, but I think it's a nice addition).

## Test Plan

Hovering over Rendered By elements successfully highlight the DOM elements.

![HighlightRenderedBy](https://user-images.githubusercontent.com/23095052/78378389-2e158b80-75d1-11ea-89d1-d6c3e743d67a.gif)

gaearon commented:
Thanks for the PR!

I should've been clearer. This might be a good change on its own (I'll defer to @bvaughn on the visual design) but what I'm looking for is **for them to get highlighted in the DOM** on hover, just like highlighting them in the tree does. This way you can move your mouse down to see which grandparent you want to stop at (because it highlights the appropriate area).

Hope this makes sense!

hristo-kanchev commented:
> Thanks for the PR!
> 
> I should've been clearer. This might be a good change on its own (I'll defer to @bvaughn on the visual design) but what I'm looking for is **for them to get highlighted in the DOM** on hover, just like highlighting them in the tree does. This way you can move your mouse down to see which grandparent you want to stop at (because it highlights the appropriate area).
> 
> Hope this makes sense!

Oh!

I completely misunderstood you. 😆  

Okay, everything is clear now. 

Closing this PR for now.

hristo-kanchev commented:
Hey @gaearon @bvaughn 

Just reopened the PR again and updated the description as well.

I've added the highlight native element support to the `rendered by` elements and I also kept the hovering state for those elements (the accidental feature 😂).

I can remove the hover styling but I think it's a nice to have. Ping me if it needs reverting and I'll update the PR right away.

Cheers!

bvaughn commented:
> * Extracted `highlightNativeElement` functionality in its own hook in order to reuse it in multiple components.
> * Replaced old highlight code in Tree view with the `useHighlightNativeElement` hook.

This PR is going to clash with #17934 but that's ... probably fine, whichever one lands first can just rebase on top of the other since the new hooks are the same.

hristo-kanchev commented:
> > * Extracted `highlightNativeElement` functionality in its own hook in order to reuse it in multiple components.
> > * Replaced old highlight code in Tree view with the `useHighlightNativeElement` hook.
> 
> This PR is going to clash with #17934 but that's ... probably fine, whichever one lands first can just rebase on top of the other since the new hooks are the same.

I'll keep an eye on that one. Thanks for linking the PR!

bvaughn commented on packages/react-devtools-shared/src/devtools/views/Components/SelectedElement.css:
```diff
@@ -107,12 +106,26 @@
   align-items: center;
-  margin-left: 0.5rem;
-  padding: 0;
+  padding-top: 0;
+  padding-right: 0;
+  padding-bottom: 0;
```
nit: I don't think these three `padding` assignments are required, since `HTMLSpanElement`s don't have any default padding.

hristo-kanchev replied:
Woopsie. Those are some leftovers from when I initially started developing the feature. Thanks for making the changes.

bvaughn commented on packages/react-devtools-shared/src/devtools/views/Components/SelectedElement.js:
```diff
@@ -532,18 +537,26 @@ function OwnerView({
 
+  const onMouseEnter = () => highlightNativeElement(id);
+
+  const onMouseLeave = () => clearHighlightNativeElement();
```
nit: This wrapper function isn't necessary :smile:

```suggestion
  const onMouseLeave = clearHighlightNativeElement;
```

hristo-kanchev replied:
We need a `facepalm` reaction on GitHub. 😄

bvaughn replied:
I'd use that one all the time :)
